### PR TITLE
Calculate next purchase feature for items

### LINKED ISSUE #10: 10. As an item, I want my estimated next purchase date to be computed at the time my purchase is recorded in the database so the app can learn how often I buy different items.
## Summary
In order to advise users about when to purchase things, the app needs to be able to calculate that guess and store it a future   date.

The `calculateEstimate` function from the `@the-collab-lab/shopping-list-utils` module will help us know how many days in the future a purchase should happen. In order to use `calculateEstimate`, we need to know how many days have passed since the time the item was last purchased.

## Acceptance criteria
- [x] When the user purchases an item, the item’s `dateNextPurchased` property is calculated using the `calculateEstimate` function and saved to the Firestore database
  - [x] `dateNextPurchased` is saved as **a date**, not a number
- [x] A `getDaysBetweenDates` function is exported from `utils/dates.js`  and imported into `api/firebase.js`
  - [x] This function takes two JavaScript Dates and return the number of days that have passed between them

### Working with Firestore’s dates
**Firestore returns dates as special objects called `Timestamps.`** You will need to convert the Timestamp to a JavaScript Date in order to work with it and complete this feature. The `Timestamp` object has a `.toDate()` method – see [the Firestore docs](https://firebase.google.com/docs/reference/js/v8/firebase.firestore.Timestamp#todate).

### Calculating the time between JavaScript dates
Comparing and manipulating JavaScript Dates is notoriously tricky. Here's a hint: all Date objects have a `getTime()` method that returns a representation of that date in milliseconds ([MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime)).

### Calculating the next purchase date
The `@the-collab-lab/shopping-list-utils` module should already be installed in your project. This module exports `calculateEstimate`.

`calculateEstimate` accepts three arguments. They are, in order:
- `previousEstimate` (a number): The last estimated purchase interval.
- `daysSinceLastPurchase` (a number): The number of days that have elapsed between the last purchased date and now. You will need to figure out how to calculate this argument!
- `totalPurchases` (a number): The total number of purchases made for the item,

Given this information, `calculateEstimate` returns **a whole number** indicating the number of days until the user is likely to make their next purchase.

### PR DESCRIPTION
**Pull request by alucernoni**
_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._

Co-authored by @yiremorlans 

## Description

This code imported existing function from npm shopping list utils package and created two helper functions to assist with date conversions for calculating the estimated date for next purchase. We chose this approach to keep our updateItems functions more pure and files less congested. We utilized the JS `.getDate()` and `.setDate()` methods to manipulate the date objects and store them as a readable string in our database. We also added the dark-mode-friendly version of the Chippy gif as a rollover from a prior issue.

## Related Issue

closes #10 

## Acceptance Criteria

 - [x] When the user purchases an item, the item’s dateNextPurchased property is calculated using the calculateEstimate function and saved to the Firestore database
 - [x] dateNextPurchased is saved as a date, not a number
 - [x] A getDaysBetweenDates function is exported from utils/dates.js and imported into api/firebase.js
 - [x] This function takes two JavaScript Dates and return the number of days that have passed between them

## Type of Changes

|     | Type                       |
| --- | -------------------------- |
|     | :bug: Bug fix              |
| ✓ | :sparkles: New feature     |
|     | :hammer: Refactoring       |
|     | :100: Add tests            |
|     | :link: Update dependencies |
|     | :scroll: Docs              |

## Testing Steps / QA Criteria

Log into Firestore. From the client-side, try checking an item on your shopping list. View the updated dateNextPurchased property for that item.

### CODE REVIEW

**Review by yiremorlans — APPROVED** [review]
I pulled the proposed changes into a test branch and Career Catalyst is fully functional again!! Awesome job going through all those file changes and identifying the slots that required renaming. The code is easy to understand and I have a better idea of how named slots are utilized in Astro.

File: src/api/firebase.js [hunk 1]
```diff
@@ -72,11 +76,41 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
...
+export async function updateItem(listId, item) {
+	const {
+		id,
+		dateCreated,
+		dateLastPurchased,
+		dateNextPurchased,
+		totalPurchases,
+	} = item;
+
+	const transformToJSDate = (date) => {
```
Comment (yiremorlans) [inline src/api/firebase.js:93]: This is a helper function takes in a Firestore Date object and returns a JS Date object that can be used.

File: src/api/firebase.js [hunk 2]
```diff
@@ -72,11 +76,41 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
 		totalPurchases: 0,
 	});
 }
+/**
+ * Update an existing item in the user's list in Firestore.
+ * @param {string} listId The id of the list we're adding to.
+ * @param {Object} item Information about the new item.
+ **/
+export async function updateItem(listId, item) {
+	const {
```
Comment (yiremorlans) [inline src/api/firebase.js:85]: The deconstructed object was moved from the component level to the Firebase API instead.

File: src/api/firebase.js [hunk 3]
```diff
@@ -72,11 +76,41 @@ export async function addItem(listId, { itemName, daysUntilNextPurchase }) {
...
+		transformToJSDate(dateNextPurchased),
+	);
+	const daysSinceLastPurchase = getDaysBetweenDates(
+		transformToJSDate(dateLastPurchased ? dateLastPurchased : dateCreated),
+		new Date(),
+	);
 
-export async function updateItem(listId, itemId) {
-	const itemRef = doc(db, listId, itemId);
+	const estimatedNextPurchase = calculateEstimate(
```
Comment (yiremorlans) [inline src/api/firebase.js:105]: The acceptance criteria require the dateNextPurchased to be a date and not a whole number returned from calculateEstimate. Therefore we need to store it in a variable to be used in getNextPurchaseDate, which returns a date.

File: src/components/ListItem.jsx [hunk 4]
```diff
@@ -4,35 +4,34 @@ import { updateItem } from '../api/firebase';
...
+		updateItem(listId, item);
 	};
-
 	useEffect(() => {
-		if (dateLastPurchased) {
-			const lastPurchasedDate = dateLastPurchased.toDate();
-			const hoursPlus24 = lastPurchasedDate.getHours() + 24;
+		if (item.dateLastPurchased) {
+			const lastPurchasedDate = item.dateLastPurchased.toDate();
+			const twentyFourHrs = 24;
```
Comment (yiremorlans) [inline src/components/ListItem.jsx:14]: We had addressed the removal of magic numbers last issue but it was accidentally refactored.

File: src/utils/dates.js [hunk 5]
```diff
@@ -10,3 +10,28 @@ const ONE_DAY_IN_MILLISECONDS = 86400000;
 export function getFutureDate(offset) {
 	return new Date(Date.now() + offset * ONE_DAY_IN_MILLISECONDS);
 }
+
+/**
+ * startDate and endDate are required to be two JavaScript Date objects in order for the compare function to work
+ * @param {Date} startDate
+ * @param {Date} endDate
+ */
+export function getDaysBetweenDates(startDate, endDate) {
```
Comment (adidalal) [inline src/utils/dates.js:19]: would love to see a couple more tests around the logic as we chatted in office hours (edge cases/boundaries, expected behavior, unexpected behavior)

File: src/views/List.jsx [hunk 6]
```diff
@@ -1,7 +1,7 @@
 import { ListItem } from '../components';
 import { useState } from 'react';
 import { useNavigate } from 'react-router-dom';
-import Chippy2 from '/img/Chippy2.gif';
+import Chippy3 from '/img/Chippy3.gif';
```
Comment (adidalal) [inline src/views/List.jsx:4]: love this versioning

File: tests/dates.test.jsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { getDaysBetweenDates } from '../src/utils';
```
Comment (adidalal) [inline tests/dates.test.jsx:1]: 💯

**Review by adidalal — APPROVED** [review]

File: src/components/ListItem.jsx (hunk 4 above)
Comment (fakehouseplant) [reply src/components/ListItem.jsx:14]: Good looking out for this.

File: src/api/firebase.js (hunk 1 above)
Comment (fakehouseplant) [reply src/api/firebase.js:93]: It feels really useful to turn this into a helper function so it can be reused.

**Review by fakehouseplant — APPROVED** [review]

File: tests/dates.test.jsx [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+		expect(getDaysBetweenDates(date1, date2)).toEqual(1);
+	});
+
+	it('given two dates in any order, then the days between returned is always positive', () => {
+		let date1 = new Date('2023-01-01');
+		let date2 = new Date('2023-01-02');
+
+		expect(getDaysBetweenDates(date2, date1)).toEqual(1);
+	});
+});
```
Comment (drakenguyen4000) [inline tests/dates.test.jsx:17]: I love that you included a test case.

**Review by drakenguyen4000 — APPROVED** [review]
I just tested it and the code created a nextDatePurchase in Firebase in required.  The code looks clean.  This looked like a lot of coding for just this feature.  Great job Annemarie and Yire!
